analytic_engine/src/sst/parquet/builder.rs

```diff
@@ -80,28 +74,78 @@ impl RecordBytesReader {
-                let buf_len = arrow_record_batch_vec.len();
-                self.fetched_row_num = 0;
-                let row_num = parquet_encoder
-                    .encode_record_batch(arrow_record_batch_vec)
-                    .map_err(|e| Box::new(e) as _)
-                    .context(EncodeRecordBatch)?;
-                arrow_record_batch_vec = Vec::with_capacity(buf_len);
-                self.total_row_num.fetch_add(row_num, Ordering::Relaxed);
+            fetched_row_num += record_batch.num_rows();
+            pending_record_batch
+                .get_or_insert_with(Default::default)
+                .push(record_batch);
+
+            // reach batch limit, append to self and reset counter and pending batch
+            if fetched_row_num >= self.num_rows_per_row_group {
+                fetched_row_num = 0;
+                self.fetched_record_batch
+                    .push(pending_record_batch.take().unwrap());
             }
         }
 
-        // final check if there is any record batch left
-        if self.fetched_row_num != 0 {
+        if let Some(remaining) = pending_record_batch.take() {
+            self.fetched_record_batch.push(remaining);
+        }
+
+        Ok(())
+    }
+
+    fn build_bloom_filter(&self) -> BloomFilter {
+        let filters = self
+            .fetched_record_batch
+            .iter()
+            .map(|row_group_batch| {
+                let mut row_group_filters =
+                    vec![Bloom::default(); row_group_batch[0].num_columns()];
+
+                for partial_batch in row_group_batch {
+                    for (col_idx, column) in partial_batch.columns().iter().enumerate() {
+                        for row in 0..column.num_rows() {
+                            let datum = column.datum(row);
+                            let bytes = datum.as_bytes();
+                            row_group_filters[col_idx].accrue(Input::Raw(&bytes));
+                        }
+                    }
+                }
+
+                row_group_filters
+            })
+            .collect::<Vec<_>>();
+
+        BloomFilter::new(filters)
+    }
+
+    async fn read_all(mut self) -> Result<Vec<u8>> {
+        self.partition_record_batch().await?;
+        let filters = self.build_bloom_filter();
+        self.meta_data.bloom_filter = filters;
+
+        let mut parquet_encoder = ParquetEncoder::try_new(
+            self.num_rows_per_row_group,
+            self.compression,
+            self.meta_data,
+        )
+        .map_err(|e| Box::new(e) as _)
+        .context(EncodeRecordBatch)?;
+
+        // process record batch stream
+        let mut arrow_record_batch_vec = Vec::new();
+        for record_batches in self.fetched_record_batch {
+            for batch in record_batches {
+                arrow_record_batch_vec.push(batch.into_record_batch().into_arrow_record_batch());
+            }
+
+            let buf_len = arrow_record_batch_vec.len();
```

Comment:
One thing I'm worried about is that the real row group distribution may not be the exactly same as the `self.fetched_record_batch`. The bloom filters for row groups will be wrong if it is indeed not what we expect.

Reply:
In current implementation, rows inside a row group(except the last one) will be larger than given `num_rows_per_row_group`.

I don't see where might be wrong this this case.

Reply:
What I mean is that one `arrow_record_batch_vec` from `partitioned_record_batch` here maybe not the real row group in the parquet, which leads to wrong usage of the bloom filters because it is built from the `partitioned_record_batch`.

Reply:
Got your point.

https://github.com/jiacai2050/arrow-rs/blob/a881083c425dfbc1a87fb510bc32644bca16fbf3/parquet/src/arrow/arrow_writer/mod.rs#L152

In `flush_completed` of parquet,  it will write one row group once `max_row_group_size` is reached, it's the same with what we expect, at least for now.

Reply:
According to https://github.com/jiacai2050/arrow-rs/blob/a881083c425dfbc1a87fb510bc32644bca16fbf3/parquet/src/arrow/arrow_writer/mod.rs#L193-L198, one row group contains at most `max_row_group_size` which is different from the distribution of our partitioned_record_batch.

Reply:
fixed